**Alumno:** César Emilio García Ávalos

**Actividad:** M3_AI3_Preparar datos de Facebook por país

In [2]:
library(readr)
library(tidyr)
library(dplyr)
install.packages("MPV")
library(MPV)
install.packages("MASS")
library(MASS)


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union


Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependency ‘randomForest’


Loading required package: lattice

Loading required package: KernSmooth

KernSmooth 2.23 loaded
Copyright M. P. Wand 1997-2009

Loading required package: randomForest

randomForest 4.7-1.1

Type rfNews() to see new features/changes/bug fixes.


Attaching package: ‘randomForest’


The following object is masked from ‘package:dplyr’:

    combine


Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)


Attaching package: ‘MASS’


The following object is masked from ‘package:MPV’:

    cement


The following object is masked from ‘package:dplyr’:

    select




In [22]:
df <- table.b3[-c(23,25),]

In [23]:
colnames(df) <- c("Miles/gallon","Displacement","Horsepower","Torque","Compression ratio","Rear axle ratio","Carburetor","No. of transmission speeds","Overall length","Width","Weight","Type of transmission")

In [24]:
summary(df)

  Miles/gallon    Displacement     Horsepower        Torque     
 Min.   :11.20   Min.   : 85.3   Min.   : 70.0   Min.   : 81.0  
 1st Qu.:16.43   1st Qu.:226.5   1st Qu.:106.0   1st Qu.:171.2  
 Median :19.30   Median :318.0   Median :141.5   Median :243.0  
 Mean   :20.04   Mean   :286.0   Mean   :137.0   Mean   :217.9  
 3rd Qu.:21.49   3rd Qu.:351.0   3rd Qu.:165.0   3rd Qu.:258.8  
 Max.   :36.50   Max.   :500.0   Max.   :223.0   Max.   :366.0  
 Compression ratio Rear axle ratio   Carburetor    No. of transmission speeds
 Min.   :8.000     Min.   :2.450   Min.   :1.000   Min.   :3.000             
 1st Qu.:8.000     1st Qu.:2.710   1st Qu.:2.000   1st Qu.:3.000             
 Median :8.325     Median :3.000   Median :2.000   Median :3.000             
 Mean   :8.313     Mean   :3.059   Mean   :2.567   Mean   :3.333             
 3rd Qu.:8.500     3rd Qu.:3.243   3rd Qu.:4.000   3rd Qu.:3.000             
 Max.   :9.000     Max.   :4.300   Max.   :4.000   Max.   :5.000             

In [25]:
sum(is.na(df))

[1] 0

In [26]:
# Calcula la matriz de correlación
correlation_matrix <- cor(df)
correlation_matrix

Miles/gallon Displacement Horsepower Torque    
Miles/gallon                1.0000000   -0.8721701   -0.7968304 -0.8495915
Displacement               -0.8721701    1.0000000    0.9408473  0.9891628
Horsepower                 -0.7968304    0.9408473    1.0000000  0.9643592
Torque                     -0.8495915    0.9891628    0.9643592  1.0000000
Compression ratio           0.4223725   -0.3469725   -0.2898995 -0.3259992
Rear axle ratio             0.6347500   -0.6720903   -0.5509642 -0.6728661
Carburetor                 -0.4718055    0.6427984    0.7614190  0.6531263
No. of transmission speeds  0.7077682   -0.7719151   -0.6259445 -0.7461800
Overall length             -0.7528208    0.8623681    0.8027387  0.8641224
Width                      -0.7629952    0.7974811    0.7105117  0.7881284
Weight                     -0.8528801    0.9515520    0.8878810  0.9434871
Type of transmission       -0.7212809    0.8244446    0.7086735  0.8012765
                           Compression ratio Rear axle ratio Carburetor 
Miles/gallon                0.42237247        0.6347500      -0.47180548
Displacement               -0.34697246       -0.6720903       0.64279836
Horsepower                 -0.28989951       -0.5509642       0.76141897
Torque                     -0.32599915       -0.6728661       0.65312630
Compression ratio           1.00000000        0.4137808       0.03748643
Rear axle ratio             0.41378081        1.0000000      -0.21952829
Carburetor                  0.03748643       -0.2195283       1.00000000
No. of transmission speeds  0.55823570        0.8717662      -0.27563863
Overall length             -0.30415026       -0.5613315       0.42206800
Width                      -0.37817358       -0.4534470       0.30038618
Weight                     -0.35845879       -0.5798617       0.52036693
Type of transmission       -0.44054570       -0.7546650       0.39548928
                           No. of transmission speeds Overall length Width     
Miles/gallon                0.7077682                 -0.7528208     -0.7629952
Displacement               -0.7719151                  0.8623681      0.7974811
Horsepower                 -0.6259445                  0.8027387      0.7105117
Torque                     -0.7461800                  0.8641224      0.7881284
Compression ratio           0.5582357                 -0.3041503     -0.3781736
Rear axle ratio             0.8717662                 -0.5613315     -0.4534470
Carburetor                 -0.2756386                  0.4220680      0.3003862
No. of transmission speeds  1.0000000                 -0.6552065     -0.6551300
Overall length             -0.6552065                  1.0000000      0.8831512
Width                      -0.6551300                  0.8831512      1.0000000
Weight                     -0.7058126                  0.9554541      0.8994711
Type of transmission       -0.8506963                  0.6824919      0.6326677
                           Weight     Type of transmission
Miles/gallon               -0.8528801 -0.7212809          
Displacement                0.9515520  0.8244446          
Horsepower                  0.8878810  0.7086735          
Torque                      0.9434871  0.8012765          
Compression ratio          -0.3584588 -0.4405457          
Rear axle ratio            -0.5798617 -0.7546650          
Carburetor                  0.5203669  0.3954893          
No. of transmission speeds -0.7058126 -0.8506963          
Overall length              0.9554541  0.6824919          
Width                       0.8994711  0.6326677          
Weight                      1.0000000  0.7530353          
Type of transmission        0.7530353  1.0000000

In [27]:
modelo1 <- lm(`Miles/gallon`~., data=df, family = gaussian)
summary(modelo1)
formula_completa = as.formula("`Miles/gallon` ~ `Compression ratio` + `No. of transmission speeds` + `Overall length` +
`Rear axle ratio` + `Type of transmission` + Carburetor + Displacement + Horsepower + Torque + Weight + Width")
modelo_vacio <- lm(`Miles/gallon`~1, data = df, family = gaussian)

Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”



Call:
lm(formula = `Miles/gallon` ~ ., data = df, family = gaussian)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.3441 -1.6711 -0.4486  1.4906  5.2508 

Coefficients:
                              Estimate Std. Error t value Pr(>|t|)  
(Intercept)                  17.339838  30.355375   0.571   0.5749  
Displacement                 -0.075588   0.056347  -1.341   0.1964  
Horsepower                   -0.069163   0.087791  -0.788   0.4411  
Torque                        0.115117   0.088113   1.306   0.2078  
`Compression ratio`           1.494737   3.101464   0.482   0.6357  
`Rear axle ratio`             5.843495   3.148438   1.856   0.0799 .
Carburetor                    0.317583   1.288967   0.246   0.8082  
`No. of transmission speeds` -3.205390   3.109185  -1.031   0.3162  
`Overall length`              0.180811   0.130301   1.388   0.1822  
Width                        -0.397945   0.323456  -1.230   0.2344  
Weight                       -0.005115   0.005896  -0.868   0.

Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”


Preliminarmente no se identifica ninguna variable estadísticamente significativa en cuanto a la explicación de la variación en la variable de respuesta "Miles/gallon", se procederá a realizar métodos de selección para simplificar el modelo

In [28]:
backward <- step(modelo1, trace = FALSE, direction = "backward", k = log(nrow(df)))

Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”
Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”
Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”
Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”
Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”
Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”
Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”
Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) 

In [29]:
backward$anova
summary(backward)

Step                           Df Deviance   Resid. Df Resid. Dev AIC     
1                                NA         NA 18        187.4007   95.77592
2 - `Type of transmission`        1  0.4648362 19        187.8655   92.44904
3 - Carburetor                    1  0.5356445 20        188.4012   89.13326
4 - `Compression ratio`           1  3.4514854 21        191.8526   86.27668
5 - Horsepower                    1 10.7796848 22        202.6323   84.51545
6 - `No. of transmission speeds`  1 10.4571693 23        213.0895   82.62383
7 - Torque                        1  4.8720101 24        217.9615   79.90082
8 - Displacement                  1  0.7654631 25        218.7270   76.60479
9 - Width                         1  5.0970905 26        223.8241   73.89468


Call:
lm(formula = `Miles/gallon` ~ `Rear axle ratio` + `Overall length` + 
    Weight, data = df, family = gaussian)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.6101 -1.9868 -0.6613  2.0369  5.8811 

Coefficients:
                   Estimate Std. Error t value Pr(>|t|)    
(Intercept)        4.590404  11.771925   0.390   0.6998    
`Rear axle ratio`  2.597240   1.264562   2.054   0.0502 .  
`Overall length`   0.217814   0.087817   2.480   0.0199 *  
Weight            -0.009485   0.001994  -4.757 6.38e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.934 on 26 degrees of freedom
Multiple R-squared:  0.8035,	Adjusted R-squared:  0.7808 
F-statistic: 35.44 on 3 and 26 DF,  p-value: 2.462e-09


In [30]:
frontward <- step(modelo_vacio, trace=FALSE, direction = "forward", k = log(nrow(df)), scope = formula_completa)

Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”


In [31]:
frontward$anova
summary(frontward)

Step           Df Deviance Resid. Df Resid. Dev AIC      
1                NA       NA 29        1139.1050  112.50522
2 + Displacement -1 866.4953 28         272.6097   73.00771


Call:
lm(formula = `Miles/gallon` ~ Displacement, data = df, family = gaussian)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.6063 -2.0276 -0.0457  1.4531  7.0213 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  33.490010   1.535476  21.811  < 2e-16 ***
Displacement -0.047026   0.004985  -9.434 3.43e-10 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.12 on 28 degrees of freedom
Multiple R-squared:  0.7607,	Adjusted R-squared:  0.7521 
F-statistic:    89 on 1 and 28 DF,  p-value: 3.429e-10


In [32]:
both <- step(modelo_vacio, trace = FALSE, direction = "both", k = log(nrow(df)), scope = formula_completa)
both$anova
summary(both)

Warning message:
“In lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) :
 extra argument ‘family’ will be disregarded”


Step           Df Deviance Resid. Df Resid. Dev AIC      
1                NA       NA 29        1139.1050  112.50522
2 + Displacement -1 866.4953 28         272.6097   73.00771


Call:
lm(formula = `Miles/gallon` ~ Displacement, data = df, family = gaussian)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.6063 -2.0276 -0.0457  1.4531  7.0213 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  33.490010   1.535476  21.811  < 2e-16 ***
Displacement -0.047026   0.004985  -9.434 3.43e-10 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.12 on 28 degrees of freedom
Multiple R-squared:  0.7607,	Adjusted R-squared:  0.7521 
F-statistic:    89 on 1 and 28 DF,  p-value: 3.429e-10


In [33]:
# Comparar los BIC
paste("BIC backward: ",BIC(backward))
paste("BIC frontward: ", BIC(frontward))
paste("BIC both :", BIC(both))

[1] "BIC backward:  162.432186920045"

[1] "BIC frontward:  161.545219767504"

[1] "BIC both : 161.545219767504"

Según los diferentes métodos de selección, el modelo con el menor BIC es el modelo reducido con frontward donde se considera únicamente la variable de desplazamiento, el coeficiente asociado al desplazamiento es aproximadamente -0.047. Esto indica que, en promedio, por cada aumento unitario en el desplazamiento (por ejemplo, un aumento de 1 pulgada cúbica en el tamaño del motor), se espera que las millas por galón disminuyan en 0.047 unidades.

In [34]:
formula_reducida = as.formula("`Miles/gallon` ~ Displacement")

In [35]:
# Función para hacer los nombres de columnas sintácticamente válidos
make_valid_names <- function(df) {
  names(df) <- make.names(names(df), unique = TRUE)
  return(df)
}

# Definir la función U
calcular_U <- function(modelo, df, response) {
  n <- nrow(df)
  Y <- df[[response]]
  Y_hat <- predict(modelo, df)

  residuo_cuadratico_medio <- mean((Y - Y_hat)^2)^0.5
  Y_cuadratico_medio <- mean(Y^2)^0.5
  Y_hat_cuadratico_medio <- mean(Y_hat^2)^0.5

  numero_variables <- length(coef(modelo)) - 1  # restar el intercepto

  U <- residuo_cuadratico_medio / (Y_cuadratico_medio + Y_hat_cuadratico_medio) + 0.05 * numero_variables
  return(U)
}

# Función para realizar STEPWISE (Forward y Backward)
stepwise_U <- function(df, response, direction = "both") {
  library(MASS)

  # Hacer los nombres de las columnas válidos
  df <- make_valid_names(df)
  response <- make.names(response)

  # Crear la fórmula de la respuesta
  response_formula <- as.formula(paste(response, "~ ."))

  # Modelo inicial: nulo para forward, completo para backward
  if (direction == "forward") {
    modelo_inicial <- lm(as.formula(paste(response, "~ 1")), data = df)
    variables_disponibles <- setdiff(names(df), response)
  } else {
    modelo_inicial <- lm(response_formula, data = df)
    variables_disponibles <- names(coef(modelo_inicial))
  }

  mejor_modelo <- modelo_inicial
  mejor_U <- calcular_U(mejor_modelo, df, response)

  mejora <- TRUE

  while (mejora) {
    mejora <- FALSE
    posibles_modelos <- list()
    posibles_U <- c()

    if (direction == "forward" || direction == "both") {
      for (var in setdiff(variables_disponibles, names(coef(mejor_modelo)))) {
        modelo <- update(mejor_modelo, paste(". ~ . +", var))
        U <- calcular_U(modelo, df, response)
        posibles_modelos <- c(posibles_modelos, list(modelo))
        posibles_U <- c(posibles_U, U)
      }
    }

    if (direction == "backward" || direction == "both") {
      for (var in names(coef(mejor_modelo))) {
        if (var != "(Intercept)") {
          modelo <- update(mejor_modelo, paste(". ~ . -", var))
          U <- calcular_U(modelo, df, response)
          posibles_modelos <- c(posibles_modelos, list(modelo))
          posibles_U <- c(posibles_U, U)
        }
      }
    }

    # Encontrar el mejor modelo de los posibles
    if (length(posibles_U) > 0) {
      mejor_posible_U <- min(posibles_U)
      if (mejor_posible_U < mejor_U) {
        mejor_U <- mejor_posible_U
        mejor_modelo <- posibles_modelos[[which.min(posibles_U)]]
        mejora <- TRUE

        if (direction == "forward" || direction == "both") {
          variables_disponibles <- setdiff(variables_disponibles, names(coef(mejor_modelo)))
        }
        if (direction == "backward" || direction == "both") {
          variables_disponibles <- names(coef(mejor_modelo))
        }
      }
    }
  }

  return(list(mejor_modelo = mejor_modelo, mejor_U = mejor_U))
}

# Cargar los datos y hacer los nombres de columnas válidos
df <- make_valid_names(df)
response <- "Miles.gallon"  # Asegúrate de que el nombre sea válido

# Aplicar STEPWISE (forward, backward y both)
resultado_forward <- stepwise_U(df, response, direction = "forward")
resultado_backward <- stepwise_U(df, response, direction = "backward")
resultado_both <- stepwise_U(df, response, direction = "both")

# Comparar los resultados
cat("U (forward):", resultado_forward$mejor_U, "\n")
cat("U (backward):", resultado_backward$mejor_U, "\n")
cat("U (both):", resultado_both$mejor_U, "\n")

summary(resultado_forward$mejor_modelo)
summary(resultado_backward$mejor_modelo)
summary(resultado_both$mejor_modelo)


U (forward): 0.1222704 
U (backward): 0.1271878 
U (both): 0.1271878 



Call:
lm(formula = Miles.gallon ~ Displacement, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.6063 -2.0276 -0.0457  1.4531  7.0213 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  33.490010   1.535476  21.811  < 2e-16 ***
Displacement -0.047026   0.004985  -9.434 3.43e-10 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.12 on 28 degrees of freedom
Multiple R-squared:  0.7607,	Adjusted R-squared:  0.7521 
F-statistic:    89 on 1 and 28 DF,  p-value: 3.429e-10



Call:
lm(formula = Miles.gallon ~ Weight, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.4913 -1.9902 -0.7168  3.3465  7.2772 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) 40.6351436  2.4591711  16.524 5.67e-16 ***
Weight      -0.0056806  0.0006572  -8.644 2.17e-09 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.33 on 28 degrees of freedom
Multiple R-squared:  0.7274,	Adjusted R-squared:  0.7177 
F-statistic: 74.72 on 1 and 28 DF,  p-value: 2.166e-09



Call:
lm(formula = Miles.gallon ~ Weight, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.4913 -1.9902 -0.7168  3.3465  7.2772 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) 40.6351436  2.4591711  16.524 5.67e-16 ***
Weight      -0.0056806  0.0006572  -8.644 2.17e-09 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.33 on 28 degrees of freedom
Multiple R-squared:  0.7274,	Adjusted R-squared:  0.7177 
F-statistic: 74.72 on 1 and 28 DF,  p-value: 2.166e-09


El método forward ha producido el modelo con el menor valor de
𝑈, sugiriendo que es el mejor en términos de balancear el ajuste del modelo y la simplicidad de acuerdo con la fórmula de 𝑈 dada, en este modelo solamente se considera la variable "Displacement", siendo esta variable estadísticamente significativa, el coeficiente asociado al desplazamiento es aproximadamente -0.047. Esto indica que, en promedio, por cada aumento unitario en el desplazamiento (por ejemplo, un aumento de 1 pulgada cúbica en el tamaño del motor), se espera que las millas por galón disminuyan en 0.047 unidades.

In [37]:
# Definir la función U con penalización variable
calcular_U <- function(modelo, df, response, penalizacion) {
  n <- nrow(df)
  Y <- df[[response]]
  Y_hat <- predict(modelo, df)

  residuo_cuadratico_medio <- mean((Y - Y_hat)^2)^0.5
  Y_cuadratico_medio <- mean(Y^2)^0.5
  Y_hat_cuadratico_medio <- mean(Y_hat^2)^0.5

  numero_variables <- length(coef(modelo)) - 1  # restar el intercepto

  U <- residuo_cuadratico_medio / (Y_cuadratico_medio + Y_hat_cuadratico_medio) + penalizacion * numero_variables
  return(U)
}

# Función para realizar STEPWISE (Forward y Backward) con penalización variable
stepwise_U <- function(df, response, direction = "both", penalizacion = 0.05) {
  library(MASS)

  # Hacer los nombres de las columnas válidos
  df <- make_valid_names(df)
  response <- make.names(response)

  # Crear la fórmula de la respuesta
  response_formula <- as.formula(paste(response, "~ ."))

  # Modelo inicial: nulo para forward, completo para backward
  if (direction == "forward") {
    modelo_inicial <- lm(as.formula(paste(response, "~ 1")), data = df)
    variables_disponibles <- setdiff(names(df), response)
  } else {
    modelo_inicial <- lm(response_formula, data = df)
    variables_disponibles <- names(coef(modelo_inicial))
  }

  mejor_modelo <- modelo_inicial
  mejor_U <- calcular_U(mejor_modelo, df, response, penalizacion)

  mejora <- TRUE

  while (mejora) {
    mejora <- FALSE
    posibles_modelos <- list()
    posibles_U <- c()

    if (direction == "forward" || direction == "both") {
      for (var in setdiff(variables_disponibles, names(coef(mejor_modelo)))) {
        modelo <- update(mejor_modelo, paste(". ~ . +", var))
        U <- calcular_U(modelo, df, response, penalizacion)
        posibles_modelos <- c(posibles_modelos, list(modelo))
        posibles_U <- c(posibles_U, U)
      }
    }

    if (direction == "backward" || direction == "both") {
      for (var in names(coef(mejor_modelo))) {
        if (var != "(Intercept)") {
          modelo <- update(mejor_modelo, paste(". ~ . -", var))
          U <- calcular_U(modelo, df, response, penalizacion)
          posibles_modelos <- c(posibles_modelos, list(modelo))
          posibles_U <- c(posibles_U, U)
        }
      }
    }

    # Encontrar el mejor modelo de los posibles
    if (length(posibles_U) > 0) {
      mejor_posible_U <- min(posibles_U)
      if (mejor_posible_U < mejor_U) {
        mejor_U <- mejor_posible_U
        mejor_modelo <- posibles_modelos[[which.min(posibles_U)]]
        mejora <- TRUE

        if (direction == "forward" || direction == "both") {
          variables_disponibles <- setdiff(variables_disponibles, names(coef(mejor_modelo)))
        }
        if (direction == "backward" || direction == "both") {
          variables_disponibles <- names(coef(mejor_modelo))
        }
      }
    }
  }

  return(list(mejor_modelo = mejor_modelo, mejor_U = mejor_U))
}

# Cargar los datos y hacer los nombres de columnas válidos
df <- make_valid_names(df)
response <- "Miles.gallon"  # Asegúrate de que el nombre sea válido

# Valores de penalización a probar
penalizaciones <- c(0.05, 0.01, 0.0001, 0.00001)

# Almacenar los resultados
resultados <- list()

# Aplicar STEPWISE (forward) para cada valor de penalización
for (pen in penalizaciones) {
  resultado <- stepwise_U(df, response, direction = "forward", penalizacion = pen)
  resultados[[paste("penalizacion", pen, sep = "_")]] <- resultado
}

# Comparar los resultados
for (pen in penalizaciones) {
  cat("Penalización:", pen, "\n")
  cat("U (forward):", resultados[[paste("penalizacion", pen, sep = "_")]]$mejor_U, "\n")
  cat("\n")
  print(summary(resultados[[paste("penalizacion", pen, sep = "_")]]$mejor_modelo))
  cat("\n")
}


Penalización: 0.05 
U (forward): 0.1222704 


Call:
lm(formula = Miles.gallon ~ Displacement, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.6063 -2.0276 -0.0457  1.4531  7.0213 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  33.490010   1.535476  21.811  < 2e-16 ***
Displacement -0.047026   0.004985  -9.434 3.43e-10 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.12 on 28 degrees of freedom
Multiple R-squared:  0.7607,	Adjusted R-squared:  0.7521 
F-statistic:    89 on 1 and 28 DF,  p-value: 3.429e-10


Penalización: 0.01 
U (forward): 0.08227042 


Call:
lm(formula = Miles.gallon ~ Displacement, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.6063 -2.0276 -0.0457  1.4531  7.0213 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  33.490010   1.535476  21.811  < 2e-16 ***
Displacement -0.047026   0.004985  -9.434 3.43e-10 ***
---

Cuando se reduce el valor de la penalización, se está disminuyendo la "multa" por tener más variables en el modelo. Esto significa que el algoritmo de selección de variables será más propenso a seleccionar modelos con un mayor número de variables predictoras, ya que la penalización por incluir más variables en el modelo es menor.

Al final, el modelo simplificado con frontward (78.08%) proporciona un mejor ajuste en comparación con el modelo simplificado con backward (75.21%), la elección entre un mejor ajuste y un menor BIC dependerá de las prioridades y del equilibrio entre la simplicidad del modelo y su capacidad para explicar los datos observados. Si es mayor el interés en la precisión de la predicción y hay datos suficientes para justificar un modelo más complejo, se podría optar por el modelo de mejor ajuste `Miles/gallon` ~ `Rear axle ratio` + `Overall length` + Weight. Sin embargo, si la simplicidad del modelo y la generalización son más importantes, se podría preferir un modelo con un menor BIC `Miles/gallon` ~ Displacement, el último modelo mencionado resulta lógico, el desplazamiento del motor es una variable importante que está relacionada con la eficiencia del combustible en los automóviles. En general, motores más grandes tienden a consumir más combustible que motores más pequeños, para efectos de simplicidad se propondría el modelo simplificado con backward.